neqo-common/src/socket.rs

```diff
@@ -0,0 +1,217 @@
+/// socket options cannot be set.
+///
+/// # Notes
+///
+/// This function binds the UDP socket to the specified local address and performs additional
+/// configuration on the socket, such as setting socket options to request TOS and TTL
+/// information for incoming packets.
+#[allow(clippy::missing_errors_doc)]
+pub fn bind(local_addr: SocketAddr) -> io::Result<UdpSocket> {
+    let socket = match UdpSocket::bind(local_addr) {
+        Err(e) => {
+            panic!("Unable to bind UDP socket: {}", e);
+        }
+        Ok(s) => {
+            // Don't let the host stack or network path fragment our IP packets
+            // (RFC9000, Section 14).
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpDontFrag, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6DontFrag, &true),
+            };
+            assert!(res.is_ok());
+            // Request IPv4 type-of-service (TOS) and IPv6 traffic class
+            // information for all incoming packets.
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpRecvTos, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6RecvTClass, &true),
+            };
+            assert!(res.is_ok());
+            // Request IPv4 time-to-live (TTL) and IPv6 hop count
+            // information for all incoming packets.
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpRecvTtl, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6RecvHopLimit, &true),
+            };
+            assert!(res.is_ok());
+            s
+        }
+    };
+    Ok(socket)
+}
+
+/// Send the UDP datagram on the specified socket.
+///
+/// # Arguments
+///
+/// * `fd` - The UDP socket to send the datagram on.
+/// * `d` - The datagram to send.
+///
+/// # Returns
+///
+/// An `io::Result` indicating whether the datagram was sent successfully.
+///
+/// # Panics
+///
+/// Panics if the send call fails.
+///
+/// # Notes
+///
+/// On non-POSIX platforms, TOS and TTL will not be sent and hence revert to the system default.
+/// TODO: Figure out how to set TOS and TTL at least on Windows.
+#[allow(clippy::missing_errors_doc)]
+pub fn emit_datagram<S: AsRawFd + FromRawFd>(socket: &S, d: &Datagram) -> io::Result<()> {
+    let sent;
+    #[cfg(any(target_os = "macos", target_os = "linux", target_os = "android"))]
+    {
+        let iov = [IoSlice::new(&d[..])];
+        let tos = i32::from(d.tos());
+        let ttl = i32::from(d.ttl());
+        let cmsgs = match d.destination() {
+            SocketAddr::V4(..) => [IpTos(&tos), IpTtl(&ttl)],
+            SocketAddr::V6(..) => [Ipv6TClass(&tos), Ipv6HopLimit(&ttl)],
+        };
+        sent = sendmsg(
+            socket.as_raw_fd(),
+            &iov,
+            &cmsgs,
+            MsgFlags::empty(),
+            Some(&SockaddrStorage::from(d.destination())),
+        )
+        .unwrap();
```

Comment:
These uses of `unwrap()` still concern me a little; convince me that this is rather than wrapping the error in a new type.  You aren't using the `io::Result::Err` branch on this either.